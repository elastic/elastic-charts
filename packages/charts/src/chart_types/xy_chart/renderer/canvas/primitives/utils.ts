/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { RGBtoString } from '../../../../../common/color_library_wrappers';
import { Fill, Stroke } from '../../../../../geoms/types';
import { MIN_STROKE_WIDTH } from './line';

/**
 * WARNING: This function modify directly, without saving, the context calling the fill() and/or stroke() if defined
 * @internal
 */
export function fillAndStroke(ctx: CanvasRenderingContext2D, fill?: Fill, stroke?: Stroke, path?: Path2D) {
  if (fill) {
    ctx.fillStyle = RGBtoString(fill.color);
    if (path) {
      ctx.fill(path);
    } else {
      ctx.fill();
    }
  }
  if (stroke && stroke.width > MIN_STROKE_WIDTH) {
    ctx.strokeStyle = RGBtoString(stroke.color);
    ctx.lineWidth = stroke.width;
    if (stroke.dash) {
      ctx.setLineDash(stroke.dash);
    }
    if (path) {
      ctx.stroke(path);
    } else {
      ctx.stroke();
    }
  }
}
