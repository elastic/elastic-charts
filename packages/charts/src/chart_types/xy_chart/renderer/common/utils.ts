/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { mergePartial } from '../../../../utils/common';
import { GeometryStateStyle, SharedGeometryStateStyle } from '../../../../utils/themes/theme';
import { TimeFunction } from '../../../../utils/time_functions';
import { AnimationOptions } from '../canvas/animations/animation';
import { AnimationConfig, AnnotationAnimationTrigger } from './../../utils/specs';

const DEFAULT_ANNOTATION_ANIMATION_OPTIONS: AnimationOptions = {
  enabled: true,
  duration: 250,
  delay: 50,
  snapValues: [],
  timeFunction: TimeFunction.easeInOut,
};

/** @internal */
export interface AnnotationHoverParams {
  style: GeometryStateStyle;
  isHighlighted: boolean;
  shouldTransition: boolean;
  options: AnimationOptions;
}

/** @internal */
export type GetAnnotationParamsFn = (id: string) => AnnotationHoverParams;

/**
 * Returns function to get geometry styles for a given id
 * @internal
 */
export const getAnnotationHoverParamsFn =
  (
    hoveredElementIds: string[],
    styles: SharedGeometryStateStyle,
    animations: AnimationConfig<AnnotationAnimationTrigger>[] = [],
  ): GetAnnotationParamsFn =>
  (id) => {
    const fadeOutConfig = animations.find(({ trigger }) => trigger === AnnotationAnimationTrigger.FadeOnFocusingOthers);
    const isHighlighted = hoveredElementIds.includes(id);
    const style =
      hoveredElementIds.length === 0 || !fadeOutConfig
        ? styles.default
        : isHighlighted
        ? styles.highlighted
        : styles.unhighlighted;
    const shouldTransition = !isHighlighted && hoveredElementIds.length > 0;

    return {
      style,
      isHighlighted,
      shouldTransition,
      options: mergePartial(DEFAULT_ANNOTATION_ANIMATION_OPTIONS, fadeOutConfig?.options),
    };
  };
