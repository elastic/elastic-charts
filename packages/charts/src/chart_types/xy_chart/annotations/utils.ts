/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { SettingsSpec } from '../../../specs';
import { Rotation, Position } from '../../../utils/common';
import { Dimensions } from '../../../utils/dimensions';
import { AnnotationId, AxisId, GroupId } from '../../../utils/ids';
import { Point } from '../../../utils/point';
import { AxisStyle } from '../../../utils/themes/theme';
import { SmallMultipleScales } from '../state/selectors/compute_small_multiple_scales';
import { getAxesSpecForSpecId } from '../state/utils/spec';
import { ComputedGeometries } from '../state/utils/types';
import { AnnotationDomainType, AnnotationSpec, AxisSpec, isLineAnnotation } from '../utils/specs';
import { computeLineAnnotationDimensions } from './line/dimensions';
import { computeRectAnnotationDimensions } from './rect/dimensions';
import { AnnotationDimensions } from './types';

/** @internal */
export function getAnnotationAxis(
  axesSpecs: AxisSpec[],
  groupId: GroupId,
  domainType: AnnotationDomainType,
  chartRotation: Rotation,
): Position | undefined {
  const { xAxis, yAxis } = getAxesSpecForSpecId(axesSpecs, groupId, chartRotation);
  const isXDomainAnnotation = isXDomain(domainType);
  return isXDomainAnnotation ? xAxis?.position : yAxis?.position;
}

/** @internal */
export function isXDomain(domainType: AnnotationDomainType): boolean {
  return domainType === AnnotationDomainType.XDomain;
}

/** @internal */
export function getTransformedCursor(
  cursorPosition: Point,
  chartDimensions: Dimensions,
  chartRotation: Rotation | null,
  /**
   * getTransformedCursor to account for projected cursor position relative to chart dimensions
   */
  projectArea = false,
): Point {
  const { height, width, left, top } = chartDimensions;
  let { x, y } = cursorPosition;

  if (projectArea) {
    x = cursorPosition.x - left;
    y = cursorPosition.y - top;
  }

  if (chartRotation === null) {
    return { x, y };
  }

  switch (chartRotation) {
    case 90:
      return { x: y, y: width - x };
    case -90:
      return { x: height - y, y: x };
    case 180:
      return { x: width - x, y: height - y };
    case 0:
    default:
      return { x, y };
  }
}

/** @internal */
export function invertTransformedCursor(
  cursorPosition: Point,
  chartDimensions: Dimensions,
  chartRotation: Rotation | null,
  /**
   * Used to account for projected cursor position relative to chart dimensions
   */
  projectArea = false,
): Point {
  const { height, width, left, top } = chartDimensions;
  let { x, y } = cursorPosition;

  switch (chartRotation) {
    case 0:
    case null:
      break;
    case 90:
      x = width - cursorPosition.y;
      y = cursorPosition.x;
      break;
    case -90:
      y = height - cursorPosition.x;
      x = cursorPosition.y;
      break;
    case 180:
    default:
      y = height - cursorPosition.y;
      x = width - cursorPosition.x;
  }

  if (projectArea) {
    x += left;
    y += top;
  }

  return { x, y };
}

/** @internal */
export function computeAnnotationDimensions(
  annotations: AnnotationSpec[],
  { rotation: chartRotation }: Pick<SettingsSpec, 'rotation'>,
  { scales: { xScale, yScales } }: Pick<ComputedGeometries, 'scales'>,
  axesSpecs: AxisSpec[],
  isHistogramModeEnabled: boolean,
  smallMultipleScales: SmallMultipleScales,
  getAxisStyle: (id?: AxisId) => AxisStyle,
): Map<AnnotationId, AnnotationDimensions> {
  return annotations.reduce<Map<AnnotationId, AnnotationDimensions>>((annotationDimensions, annotationSpec) => {
    const { id } = annotationSpec;

    if (isLineAnnotation(annotationSpec)) {
      const { groupId, domainType } = annotationSpec;
      const annotationAxisPosition = getAnnotationAxis(axesSpecs, groupId, domainType, chartRotation);

      const dimensions = computeLineAnnotationDimensions(
        annotationSpec,
        chartRotation,
        yScales,
        xScale,
        smallMultipleScales,
        isHistogramModeEnabled,
        annotationAxisPosition,
      );
      if (dimensions) {
        annotationDimensions.set(id, dimensions);
      }
      return annotationDimensions;
    } else {
      const dimensions = computeRectAnnotationDimensions(
        annotationSpec,
        yScales,
        xScale,
        axesSpecs,
        smallMultipleScales,
        chartRotation,
        getAxisStyle,
        isHistogramModeEnabled,
      );

      if (dimensions) {
        annotationDimensions.set(id, dimensions);
      }
      return annotationDimensions;
    }
  }, new Map());
}
