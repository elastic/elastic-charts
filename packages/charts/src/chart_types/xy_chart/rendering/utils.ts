/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { LegendItem } from '../../../common/legend';
import { ScaleBand, ScaleContinuous } from '../../../scales';
import { isLogarithmicScale } from '../../../scales/types';
import { MarkBuffer } from '../../../specs';
import { getDistance } from '../../../utils/common';
import { BarGeometry, ClippedRanges, isPointGeometry, PointGeometry } from '../../../utils/geometry';
import { GeometryStateStyle, SharedGeometryStateStyle } from '../../../utils/themes/theme';
import { DataSeriesDatum, FilledValues, XYChartSeriesIdentifier } from '../utils/series';

/** @internal */
export interface MarkSizeOptions {
  enabled: boolean;
  ratio?: number;
}

/**
 * Returns value of `y1` or `filled.y1` or null by default.
 * Passing a filled key (x, y1, y0) it will return that value or the filled one
 * @internal
 */
export function getYDatumValueFn(valueName: keyof Omit<FilledValues, 'x'> = 'y1') {
  return (datum: DataSeriesDatum, returnFilled = true): number | null => {
    const value = datum[valueName];
    if (value !== null || !returnFilled) {
      return value;
    }
    return datum.filled?.[valueName] ?? null;
  };
}

/**
 *
 * @param param0
 * @internal
 */
export function isDatumFilled({ filled, initialY1 }: DataSeriesDatum) {
  return filled?.x !== undefined || filled?.y1 !== undefined || initialY1 === null || initialY1 === undefined;
}

/**
 * Gets clipped ranges that have been fitted to values
 * @param dataset
 * @param xScale
 * @param xScaleOffset
 * @internal
 */
export function getClippedRanges(
  dataset: DataSeriesDatum[],
  xScale: ScaleBand | ScaleContinuous,
  xScaleOffset: number,
): ClippedRanges {
  let firstNonNullX: number | null = null;
  let hasNull = false;

  const completeDatasetIsNull = dataset.every((datum) => isDatumFilled(datum));

  if (completeDatasetIsNull) return [[xScale.range[0], xScale.range[1]]];

  return dataset.reduce<ClippedRanges>((acc, data) => {
    const xScaled = xScale.scale(data.x);
    if (Number.isNaN(xScaled)) return acc;

    const xValue = xScaled - xScaleOffset + xScale.bandwidth / 2;

    if (isDatumFilled(data)) {
      const endXValue = xScale.range[1] - xScale.bandwidth * (2 / 3);
      if (firstNonNullX !== null && xValue === endXValue) {
        acc.push([firstNonNullX, xValue]);
      }
      hasNull = true;
    } else {
      if (hasNull) {
        if (firstNonNullX !== null) {
          acc.push([firstNonNullX, xValue]);
        } else {
          acc.push([0, xValue]);
        }
        hasNull = false;
      }

      firstNonNullX = xValue;
    }
    return acc;
  }, []);
}

/** @internal */
export function getGeometryStateStyle(
  seriesIdentifier: XYChartSeriesIdentifier,
  sharedGeometryStyle: SharedGeometryStateStyle,
  highlightedLegendItem?: LegendItem,
): GeometryStateStyle {
  const { default: defaultStyles, highlighted, unhighlighted } = sharedGeometryStyle;

  if (highlightedLegendItem) {
    const isPartOfHighlightedSeries = highlightedLegendItem.seriesIdentifiers.some(
      ({ key }) => key === seriesIdentifier.key,
    );

    return isPartOfHighlightedSeries ? highlighted : unhighlighted;
  }

  return defaultStyles;
}

/** @internal */
export function isPointOnGeometry(
  xCoordinate: number,
  yCoordinate: number,
  indexedGeometry: BarGeometry | PointGeometry,
  buffer: MarkBuffer,
) {
  const { x, y, transform } = indexedGeometry;
  if (isPointGeometry(indexedGeometry)) {
    const { radius } = indexedGeometry;
    const distance = getDistance(
      {
        x: xCoordinate,
        y: yCoordinate,
      },
      {
        x: x + transform.x,
        y: y + transform.y,
      },
    );

    const radiusBuffer = typeof buffer === 'number' ? buffer : buffer(radius);

    return distance <= radius + radiusBuffer;
  }
  const { width, height } = indexedGeometry;
  return yCoordinate >= y && yCoordinate <= y + height && xCoordinate >= x && xCoordinate <= x + width;
}

const getScaleTypeValueValidator = (yScale: ScaleContinuous): ((n: number) => boolean) => {
  if (!isLogarithmicScale(yScale)) return () => true;
  const domainPolarity = getDomainPolarity(yScale.domain);
  return (yValue: number) => domainPolarity === Math.sign(yValue);
};

/**
 * The default zero baseline for area charts.
 */
const DEFAULT_ZERO_BASELINE = 0;

/** @internal */
export type YDefinedFn = (datum: DataSeriesDatum, getValueAccessor: (d: DataSeriesDatum) => number | null) => boolean;

/** @internal */
export function isYValueDefinedFn(yScale: ScaleContinuous, xScale: ScaleBand | ScaleContinuous): YDefinedFn {
  const validator = getScaleTypeValueValidator(yScale);
  return (datum, getValueAccessor) => {
    const yValue = getValueAccessor(datum);
    return yValue !== null && validator(yValue) && xScale.isValueInDomain(datum.x);
  };
}

/** @internal */
export const CHROME_PINCH_BUG_EPSILON = 0.5;

/**
 * Temporary fix for Chromium bug
 * Shift a small pixel value when pixel diff is <= 0.5px
 * https://github.com/elastic/elastic-charts/issues/1053
 * https://bugs.chromium.org/p/chromium/issues/detail?id=1163912
 */
function chromeRenderBugBuffer(y1: number, y0: number): number {
  return Math.abs(y1 - y0) <= CHROME_PINCH_BUG_EPSILON ? 0.5 : 0;
}

/** @internal */
export function getY1ScaledValueFn(yScale: ScaleContinuous): (datum: DataSeriesDatum) => number {
  const datumAccessor = getYDatumValueFn();
  const scaleY0Value = getY0ScaledValueFn(yScale);
  return (datum) => {
    const y1Value = yScale.scale(datumAccessor(datum));
    const y0Value = scaleY0Value(datum);
    return y1Value - chromeRenderBugBuffer(y1Value, y0Value);
  };
}

/** @internal */
export function getY0ScaledValueFn(yScale: ScaleContinuous): (datum: DataSeriesDatum) => number {
  const domainPolarity = getDomainPolarity(yScale.domain);
  const logBaseline = domainPolarity >= 0 ? Math.min(...yScale.domain) : Math.max(...yScale.domain);
  return ({ y0 }) =>
    isLogarithmicScale(yScale) // checking wrong y0 polarity
      ? y0 === null || domainPolarity !== Math.sign(y0) // if all positive domain use 1 as baseline, -1 otherwise
        ? yScale.scale(logBaseline)
        : yScale.scale(y0) // if negative value, use -1 as max reference, 1 otherwise
      : yScale.scale(y0 === null ? DEFAULT_ZERO_BASELINE : y0);
}

function getDomainPolarity(domain: number[]): number {
  // 1 if both numbers are positive, -1 if both are negative, 0 if zeros or mixed
  return Math.sign(Math.sign(domain[0]) + Math.sign(domain[1]));
}
