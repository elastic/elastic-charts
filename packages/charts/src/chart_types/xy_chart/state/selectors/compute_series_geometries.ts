/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { createCustomCachedSelector } from '../../../../state/create_selector';
import { getChartThemeSelector } from '../../../../state/selectors/get_chart_theme';
import { getSettingsSpecSelector } from '../../../../state/selectors/get_settings_spec';
import { withTextMeasure } from '../../../../utils/bbox/canvas_text_bbox_calculator';
import { computeSeriesGeometries } from '../utils/utils';
import { getFallBackTickFormatter } from './compute_axis_ticks_dimensions';
import { computeSeriesDomainsSelector } from './compute_series_domains';
import { computeSmallMultipleScalesSelector } from './compute_small_multiple_scales';
import { getSeriesColorsSelector } from './get_series_color_map';
import { getSeriesSpecsSelector, getAxisSpecsSelector } from './get_specs';
import { isHistogramModeEnabledSelector } from './is_histogram_mode_enabled';

/** @internal */
export const computeSeriesGeometriesSelector = createCustomCachedSelector(
  [
    getSeriesSpecsSelector,
    computeSeriesDomainsSelector,
    getSeriesColorsSelector,
    getChartThemeSelector,
    getSettingsSpecSelector,
    getAxisSpecsSelector,
    computeSmallMultipleScalesSelector,
    isHistogramModeEnabledSelector,
    getFallBackTickFormatter,
  ],
  (specs, domain, colors, theme, settings, axis, smScales, isHistogram, fallbackFormatter) => {
    return withTextMeasure((measureText) =>
      computeSeriesGeometries(
        specs,
        domain,
        colors,
        theme,
        settings,
        axis,
        smScales,
        isHistogram,
        fallbackFormatter,
        measureText,
      ),
    );
  },
);
