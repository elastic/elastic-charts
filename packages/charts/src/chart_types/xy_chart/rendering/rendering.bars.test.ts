/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { MockGlobalSpec, MockSeriesSpec } from '../../../mocks/specs';
import { MockStore } from '../../../mocks/store';
import { ScaleType } from '../../../scales/constants';
import { identity } from '../../../utils/common';
import { computeSeriesGeometriesSelector } from '../state/selectors/compute_series_geometries';

const SPEC_ID = 'spec_1';
const GROUP_ID = 'group_1';

describe('Rendering bars', () => {
  test('Can render two bars within domain', () => {
    const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
    const spec = MockSeriesSpec.bar({
      id: SPEC_ID,
      groupId: GROUP_ID,
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
      xAccessor: 0,
      yAccessors: [1],
      data: [
        [-200, 0],
        [0, 10],
        [1, 5],
      ], // first datum should be skipped as it's out of domain
    });
    MockStore.addSpecs(
      [spec, MockGlobalSpec.settingsNoMargins({ xDomain: [0, 1], theme: { colors: { vizColors: ['red'] } } })],
      store,
    );
    const { geometries } = computeSeriesGeometriesSelector(store.getState());

    expect(geometries.bars[0].value).toMatchSnapshot();
  });

  describe('Single series bar chart - ordinal', () => {
    test('Can render bars with value labels', () => {
      const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
      MockStore.addSpecs(
        [
          MockSeriesSpec.bar({
            id: SPEC_ID,
            groupId: GROUP_ID,
            xScaleType: ScaleType.Ordinal,
            yScaleType: ScaleType.Linear,
            xAccessor: 0,
            yAccessors: [1],
            data: [
              [-200, 0],
              [0, 10],
              [1, 5],
            ], // first datum should be skipped as it's out of domain
            displayValueSettings: {
              showValueLabel: true,
              isAlternatingValueLabel: true,
              valueFormatter: identity,
            },
          }),
          MockGlobalSpec.settingsNoMargins({ xDomain: [0, 1], theme: { colors: { vizColors: ['red'] } } }),
        ],
        store,
      );
      const { geometries } = computeSeriesGeometriesSelector(store.getState());
      expect(geometries.bars[0].value[0].displayValue).toBeDefined();
    });

    test('Can hide value labels if no formatter or showValueLabels is false/undefined', () => {
      const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
      MockStore.addSpecs(
        [
          MockSeriesSpec.bar({
            id: SPEC_ID,
            groupId: GROUP_ID,
            xScaleType: ScaleType.Ordinal,
            yScaleType: ScaleType.Linear,
            xAccessor: 0,
            yAccessors: [1],
            data: [
              [-200, 0],
              [0, 10],
              [1, 5],
            ], // first datum should be skipped as it's out of domain
            displayValueSettings: {
              showValueLabel: false,
              isAlternatingValueLabel: true,
              valueFormatter: identity,
            },
          }),
          MockGlobalSpec.settingsNoMargins({ xDomain: [0, 1], theme: { colors: { vizColors: ['red'] } } }),
        ],
        store,
      );
      const { geometries } = computeSeriesGeometriesSelector(store.getState());
      expect(geometries.bars[0].value[0].displayValue).toBeUndefined();
    });

    test('Can render bars with alternating value labels', () => {
      const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
      MockStore.addSpecs(
        [
          MockSeriesSpec.bar({
            id: SPEC_ID,
            groupId: GROUP_ID,
            xScaleType: ScaleType.Ordinal,
            yScaleType: ScaleType.Linear,
            xAccessor: 0,
            yAccessors: [1],
            data: [
              [-200, 0],
              [0, 10],
              [1, 5],
            ], // first datum should be skipped as it's out of domain
            displayValueSettings: {
              showValueLabel: true,
              isAlternatingValueLabel: true,
              valueFormatter: identity,
            },
          }),
          MockGlobalSpec.settingsNoMargins({ xDomain: [0, 1], theme: { colors: { vizColors: ['red'] } } }),
        ],
        store,
      );
      const { geometries } = computeSeriesGeometriesSelector(store.getState());

      expect(geometries.bars[0].value[0].displayValue?.text).toBeDefined();
      expect(geometries.bars[0].value[1].displayValue?.text).toBeUndefined();
    });

    test('Can render bars with contained value labels', () => {
      const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
      MockStore.addSpecs(
        [
          MockSeriesSpec.bar({
            id: SPEC_ID,
            groupId: GROUP_ID,
            xScaleType: ScaleType.Ordinal,
            yScaleType: ScaleType.Linear,
            xAccessor: 0,
            yAccessors: [1],
            data: [
              [-200, 0],
              [0, 10],
              [1, 5],
            ], // first datum should be skipped as it's out of domain
            displayValueSettings: {
              showValueLabel: true,
              isValueContainedInElement: true,
              valueFormatter: identity,
            },
          }),
          MockGlobalSpec.settingsNoMargins({ xDomain: [0, 1], theme: { colors: { vizColors: ['red'] } } }),
        ],
        store,
      );
      const { geometries } = computeSeriesGeometriesSelector(store.getState());

      expect(geometries.bars[0].value[0].displayValue?.width).toBe(50);
    });
  });
  describe('Multi series bar chart - ordinal', () => {
    const spec1Id = 'bar1';
    const spec2Id = 'bar2';
    const barSeriesSpec1 = MockSeriesSpec.bar({
      id: spec1Id,
      groupId: GROUP_ID,
      data: [
        [0, 10],
        [1, 5],
      ],
      xAccessor: 0,
      yAccessors: [1],
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
    });
    const barSeriesSpec2 = MockSeriesSpec.bar({
      id: spec2Id,
      groupId: GROUP_ID,
      data: [
        [0, 20],
        [1, 10],
      ],
      xAccessor: 0,
      yAccessors: [1],
      xScaleType: ScaleType.Ordinal,
      yScaleType: ScaleType.Linear,
    });
    const store = MockStore.default({ width: 100, height: 100, top: 0, left: 0 });
    MockStore.addSpecs(
      [
        barSeriesSpec1,
        barSeriesSpec2,
        MockGlobalSpec.settingsNoMargins({ theme: { colors: { vizColors: ['red', 'blue'] } } }),
      ],
      store,
    );

    const {
      geometries: { bars },
    } = computeSeriesGeometriesSelector(store.getState());

    test('can render first spec bars', () => {
      expect(bars[0].value).toMatchSnapshot();
    });
    test('can render second spec bars', () => {
      expect(bars[1].value).toMatchSnapshot();
    });
  });
});
