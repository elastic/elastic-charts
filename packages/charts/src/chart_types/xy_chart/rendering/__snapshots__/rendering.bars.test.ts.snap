// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering bars Can render two bars within domain 1`] = `
Array [
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 100,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 10,
    },
    "width": 50,
    "x": 0,
    "y": 0,
  },
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 50,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{spec_1}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "spec_1",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        1,
        5,
      ],
      "mark": null,
      "x": 1,
      "y": 5,
    },
    "width": 50,
    "x": 50,
    "y": 50,
  },
]
`;

exports[`Rendering bars Multi series bar chart - ordinal can render first spec bars 1`] = `
Array [
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 50,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{bar1}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "bar1",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        10,
      ],
      "mark": null,
      "x": 0,
      "y": 10,
    },
    "width": 25,
    "x": 0,
    "y": 50,
  },
  Object {
    "color": "red",
    "displayValue": undefined,
    "height": 25,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{bar1}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "bar1",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        1,
        5,
      ],
      "mark": null,
      "x": 1,
      "y": 5,
    },
    "width": 25,
    "x": 50,
    "y": 75,
  },
]
`;

exports[`Rendering bars Multi series bar chart - ordinal can render second spec bars 1`] = `
Array [
  Object {
    "color": "blue",
    "displayValue": undefined,
    "height": 100,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{bar2}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "bar2",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        0,
        20,
      ],
      "mark": null,
      "x": 0,
      "y": 20,
    },
    "width": 25,
    "x": 25,
    "y": 0,
  },
  Object {
    "color": "blue",
    "displayValue": undefined,
    "height": 50,
    "panel": Object {
      "height": 100,
      "left": 0,
      "top": 0,
      "width": 100,
    },
    "seriesIdentifier": Object {
      "key": "groupId{group_1}spec{bar2}yAccessor{1}splitAccessors{}",
      "seriesKeys": Array [
        1,
      ],
      "smHorizontalAccessorValue": undefined,
      "smVerticalAccessorValue": undefined,
      "specId": "bar2",
      "splitAccessors": Map {},
      "yAccessor": 1,
    },
    "seriesStyle": Object {
      "displayValue": Object {
        "fill": "#777",
        "fontFamily": "sans-serif",
        "fontSize": 8,
        "fontStyle": "normal",
        "offsetX": 0,
        "offsetY": 0,
        "padding": 0,
      },
      "rect": Object {
        "opacity": 1,
      },
      "rectBorder": Object {
        "strokeWidth": 1,
        "visible": false,
      },
    },
    "transform": Object {
      "x": 0,
      "y": 0,
    },
    "value": Object {
      "accessor": "y1",
      "datum": Array [
        1,
        10,
      ],
      "mark": null,
      "x": 1,
      "y": 10,
    },
    "width": 25,
    "x": 75,
    "y": 50,
  },
]
`;
